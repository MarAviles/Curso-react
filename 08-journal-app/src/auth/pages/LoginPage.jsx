import { useDispatch, useSelector } from 'react-redux';
import { useMemo } from 'react';
import { Link as RouterLink } from 'react-router-dom';
import { Grid, Typography, TextField, Button, Link } from "@mui/material"
import { Google } from "@mui/icons-material"

import { AuthLayout } from '../layout/AuthLayout'
import { useForm } from '../../hooks'
import { checkingAuthetication, startGoogleSignIn } from '../../store';

export const LoginPage = () => {

  const { status } = useSelector( state => state.auth);

  const dispatch = useDispatch();

  const { email, password, onInputChange} = useForm({
    email: 'mariana.aviles@gmail.com',
    password: '123345'
  });

  const isAuthenticating = useMemo( () =>  status == 'checking', [status]);

  const onSubmit = ( event ) => {
    event.preventDefault();
    console.log({ email, password });

    dispatch( checkingAuthetication() );
  }

  const onGoogleSignIn = () => {
    dispatch( startGoogleSignIn() );
  }

  return (
    <AuthLayout title='Login'>
      <form onSubmit={ onSubmit }>
        <Grid container>
          <Grid item xs={ 12 } sx={{ mt: 2 }}>
            <TextField 
              label="Correo" 
              type='email' 
              placeholder='correo@google.com'
              fullWidth
              name="email"
              value={ email }
              onChange={ onInputChange }
            />
          </Grid>

          <Grid item xs={ 12 } sx={{ mt: 2 }}>
            <TextField 
              label="ContraseÃ±a" 
              type='password' 
              placeholder='*********'
              fullWidth
              name="password"
              value={ password }
              onChange={ onInputChange }
            />
          </Grid>

        </Grid>

        <Grid container spacing={ 2 } sx={{ mb: 2, mt: 2 }}>
          <Grid item xs={ 12 } sm={ 6 }>
            <Button 
              disabled={ isAuthenticating }
              type='submit'
              variant='contained' 
              fullWidth>
              Login
            </Button>
          </Grid>

          <Grid item xs={ 12 } sm={ 6 }>
            <Button
              disabled={ isAuthenticating }
              variant='contained'
              fullWidth
              onClick={ onGoogleSignIn }>
              <Google/>
              <Typography sx={{ ml: 1 }}>Google</Typography>
            </Button>
          </Grid>
        </Grid>

        <Grid container direction='row' justifyContent='end'>
          <Link component={ RouterLink } color="inherit" to="/auth/register">
            Crear una cuenta
          </Link>
        </Grid>
      </form>
    </AuthLayout>
  )
}

