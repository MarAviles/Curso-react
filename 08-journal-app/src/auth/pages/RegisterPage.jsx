import { Link as RouterLink } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux'
import { Button, Grid, TextField, Typography, Link, Alert } from "@mui/material";
import { AuthLayout } from "../layout/AuthLayout"
import { useForm } from '../../hooks';
import { useState } from 'react';
import { startCreatingUserWithEmailPassword } from '../../store/thunks';
import { useMemo } from 'react';

const formData = {
  email: '',
  password: '',
  displayName: ''
}

const formValidations = {
  email: [ (value) => value.includes('@') , 'El email debe de tener un @'],
  password: [ (value) => value.length >= 6, 'El password debe de tener más de 6 letras'],
  displayName: [ (value) => value.length >= 1, 'El nombre es obligatorio']
}

export const RegisterPage = () => {

  const dispatch = useDispatch();
  const [formSubmmitted, setFormSubmmitted] = useState(false);

  const { status, errorMessage } = useSelector( state => state.auth );
  const isCheckingAuthentication = useMemo( () => status === 'checking', [status]);

  const { formState, displayName, email, password, onInputChange, 
          isFormValid, displayNameValid, emailValid, passwordValid
        } = useForm(formData, formValidations);

  const onSubmit = ( event ) => {
    event.preventDefault();
    setFormSubmmitted(true);
    if ( !isFormValid ) return;
  
    console.log('formState', formState);
    dispatch( startCreatingUserWithEmailPassword(formState));
  }

  return (
    <AuthLayout title='Crear cuenta'>
      <p>FormValid { isFormValid ? 'Valido' : 'Incorrecto'} </p>
      <form>
        <Grid container>
          <Grid item xs={ 12 } sx={{ mt: 2 }}>
              <TextField 
                label="Nombre Completo" 
                type='text' 
                placeholder='Nombre completo'
                fullWidth
                name='displayName'
                value={ displayName }
                onChange={ onInputChange }
                error= { displayNameValid && formSubmmitted }
                helperText= { displayNameValid }
              />
            </Grid>
          <Grid item xs={ 12 } sx={{ mt: 2 }}>
            <TextField 
              label="Correo" 
              type='email' 
              placeholder='correo@google.com'
              fullWidth
              name='email'
              value={ email }
              onChange={ onInputChange }
              error= { emailValid && formSubmmitted }
              helperText= { emailValid }
            />
          </Grid>

          <Grid item xs={ 12 } sx={{ mt: 2 }}>
            <TextField 
              label="Contraseña" 
              type='password' 
              placeholder='*********'
              fullWidth
              name='password'
              value={ password }
              onChange={ onInputChange }
              error= { passwordValid && formSubmmitted }
              helperText= { passwordValid }
            />
          </Grid>
        </Grid>

        <Grid container spacing={ 2 } sx={{ mb: 2, mt: 1 }}>

          <Grid 
            item 
            xs={ 12 }
            display={ !!errorMessage ? '': 'none' }
            >
            <Alert severity='error'>{errorMessage}</Alert>
          </Grid>

          <Grid item xs={ 12 }>
            <Button
              disabled={ isCheckingAuthentication }
              type='submit'
              variant='contained' 
              fullWidth>
              Crear cuenta
            </Button>
          </Grid>
        </Grid>

        <Grid container direction='row' justifyContent='end'>
          <Typography sx={{ mr: 1 }}>¿Ya tienes cuenta?</Typography>
          <Link component={ RouterLink } color="inherit" to="/auth/login">
            Ingresar
          </Link>
        </Grid>
      </form>
    </AuthLayout>
  )
}
