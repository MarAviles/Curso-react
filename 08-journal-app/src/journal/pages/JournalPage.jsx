import { Typography } from '@mui/material'
import { JornualLayout } from '../layout/JornualLayout'

export const JournalPage = () => {
  return (
    <JornualLayout>
      <Typography>Nulla anim nostrud anim mollit sunt excepteur deserunt id.</Typography>

      {/* NothingSelected */}
      {/* ViewPage */}
    </JornualLayout>
  )
}
