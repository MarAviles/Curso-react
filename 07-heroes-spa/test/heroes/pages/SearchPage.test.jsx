import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/herores/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));

beforeEach(() => jest.clearAllMocks());


describe('Prueba en el componente <SearchPage/>', () => {

    test('debe de mostrarse correctamente con valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchPage/>
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();
    });

    test('debe de mostrar a Batman y el input con el valor del queryString', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=Batman']}>
                <SearchPage/>
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('Batman');

        const img = screen.getByRole('img');
        expect( img.src ).toContain('/asset/heroes/dc-batman.jpg');

        const showSearch = screen.getByLabelText('showSearch');
        expect( showSearch.style.display ).toBe('none');
    });

    test('debe de mostrar un error si no se encuentra el hero', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=Pikachu']}>
                <SearchPage/>
            </MemoryRouter>
        );
        
        const showError = screen.getByLabelText('showError');
        expect( showError.style.display ).toBe('');

    });


    test('debe de llamar al navigate a la pantalla nueva', () => {

        const inputValue = 'Batman'

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage/>
            </MemoryRouter>
        ); 

        const input = screen.getByRole('textbox');
        const form = screen.getByLabelText('form');

        fireEvent.input( input, { target: { value: inputValue}});
        fireEvent.submit( form );

        expect( mockedUseNavigate ).toHaveBeenCalledWith(`?q=${inputValue}`);
        // expect( mockedUseNavigate ).toHaveBeenCalledWith()
    })
});