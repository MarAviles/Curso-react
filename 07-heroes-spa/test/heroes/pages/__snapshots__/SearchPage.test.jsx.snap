// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en el componente <SearchPage/> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <form>
        <input
          aria-label="form"
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="showSearch"
        class="alert alert-primary"
      >
        Search a hero
      </div>
      <div
        aria-label="showError"
        class="alert alert-danger"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
